### PR TITLE
Adding OAR executor

### PR DESCRIPTION
**Pull request by valleema**
Hello,

Based on the work of @Tintest, I made a fully functionning OAR scheduler handler for Nextflow.

I discussed with him before getting in it, and I made sure to include some of the constructive criticism found on the old pages:

https://github.com/nextflow-io/nextflow/issues/708
https://github.com/nextflow-io/nextflow/pull/766

I hope it will be of some help to others.

### CODE REVIEW

**Comment by valleema** [comment]
Hi Paolo,

I have:

- Signed-off with DCO
- Added the documentation
- Removed the Eclipse-related files

I am a little lost as to solve the Const.groovy conflict. Should I use `git revert` to the specific commit and `git checkout` the Const.groovy file? Will this work?

**Comment by pditommaso** [comment]
Hi Maxime, I think the simplest way is to revert the changes in the Const file manually (no using the `revert` command) and commit the previous version again. Alternatively, you can reset (soft) all your changes, commit and push (force) again.

One more thing, we definitively need a test class for the Oar executor. Please check the same for [SGE](https://github.com/nextflow-io/nextflow/blob/29353c901b2bee257daa4855ca921882643b52f7/modules/nextflow/src/test/groovy/nextflow/executor/SgeExecutorTest.groovy#L28) or [SLURM](https://github.com/nextflow-io/nextflow/blob/29353c901b2bee257daa4855ca921882643b52f7/modules/nextflow/src/test/groovy/nextflow/executor/SlurmExecutorTest.groovy#L28)

**Comment by valleema** [comment]
Ok I manually reverted the changes. 

Wow the test class is one long script!! I will try to make mimicking the other scheduler ones.

**Comment by pditommaso** [comment]
> Wow the test class is one long script!!

😂 

Yes, sorry, that's important to make it stable and above all to avoid that changes overtime break the existing code.

**Comment by valleema** [comment]
Yes, I was wondering If I should ask you about a local test.
Thanks!



De : Paolo Di Tommaso [mailto:notifications@github.com]
Envoyé : jeudi 24 octobre 2019 11:33
À : nextflow-io/nextflow <nextflow@noreply.github.com>
Cc : VALLEE, Maxime <maxime.vallee@chu-lyon.fr>; Author <author@noreply.github.com>
Objet : [NEWS] Re: [nextflow-io/nextflow] Adding OAR executor (#1343)


Guess you are struggling to make tests pass. Run make test locally before pushing, it's much simpler and quicker.

**Comment by valleema** [comment]
Hi Paolo, 

lucky we have a local test, I am at around few billions of those.

I have gone nearly mad, but I saw a strange behaviour, as some exactly identical tests where sometimes passing, sometimes, not.

Thanks to `cat -A` I saw that, if one mixes up tab and quadruple spaces indentations in test headers (which is obviously obvious once I copy-pasted in here... vim and VSCode were not showing the diff);

```
        when:
        task.config = new TaskConfig()
        task.config.time = '1h'
        then:
        executor.getHeaders(task) == '''
                #OAR -d /work/path
                #OAR -n nf-the_task_name
                #OAR -O /work/path/.command.out
                #OAR -E /work/path/.command.err
				#OAR -l walltime=01:00:00
                '''
                .stripIndent().leftTrim()
```

However, if the test header is fully made with tab indentation OR fully made with quadruple space indentation, the tests pass.

**Comment by pditommaso** [comment]
Well, you know tabs .. 😂 

https://www.youtube.com/watch?v=SsoOG6ZeyUI

**Comment by valleema** [comment]
Yeah, it is a good one... I will not be starting a fight over this PR. Lost few hours, but now my VSCode prefs are changed, amd I see everything 🤣 

Yoohoo Travis CI went through!!

File: modules/nextflow/src/main/groovy/nextflow/executor/OarExecutor.groovy [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        result << '-E' << quote(task.workDir.resolve(TaskRun.CMD_ERRFILE))     
+
+        if( task.config.getMemory() ) {
+            if( task.config.getMemory().toMega() < 1024 ) {
+				result << "-p" << "\"memnode=1\""
+			}
+			else {
+				result << "-p" << "\"memnode=${task.config.getMemory().toGiga().toString()}\""
+			}
+        }
```
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/executor/OarExecutor.groovy]: The indentation can be improved, here

File: modules/nextflow/src/main/groovy/nextflow/executor/OarExecutor.groovy [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        if( task.config.cpus > 1) {
+            if( task.config.time ) {
+				// cpu + time set
+				result << "-l" << "/nodes=1/core=${task.config.cpus.toString()},walltime=${task.config.getTime().format('HH:mm:ss')}"
+        	}
+			else {
+				// just cpu set
+				result << "-l" << "/nodes=1/core=${task.config.cpus.toString()}"
+			}
+		}
```
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/executor/OarExecutor.groovy]: here

File: modules/nextflow/src/main/groovy/nextflow/executor/OarExecutor.groovy [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+     *
+     * @param task The {@link TaskRun} instance to submit for execution to the cluster
+     * @param scriptFile The file containing the job launcher script
+     * @return A list representing the submit command line
+     */
+    @Override
+    List<String> getSubmitCommandLine(TaskRun task, Path scriptFile ) {
+		// Scripts need to be executable
+		scriptFile.setPermissions(7,0,0)
+        return ["oarsub", "-S", "./${scriptFile.getName()}"]
```
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/executor/OarExecutor.groovy]: here as well

File: modules/nextflow/src/main/groovy/nextflow/executor/OarExecutor.groovy [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    /**
+     * Parse the string returned by the {@code oarsub} command and extract the job ID string
+     *
+     * @param text The string returned when submitting the job
+     * @return The actual job ID string
+     */
+    static private Pattern SUBMIT_REGEX = ~/OAR_JOB_ID=(\d+)/
+	
+	@Override
+    def parseJobId(String text) {
```
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/executor/OarExecutor.groovy]: also here

File: modules/nextflow/src/main/groovy/nextflow/executor/OarExecutor.groovy [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+    @Override
+    protected List<String> getKillCommand() { ['oardel'] }
+
+    @Override
+    protected List<String> queueStatusCommand(Object queue) {
+        // To have a parsable list of jobs in queue by user
+		// see page 21 http://oar.imag.fr/docs/2.5/OAR-Documentation.pdf
+		String cmd = 'oarstat -f'
+        if( queue ) cmd += ' ' + queue
```
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/executor/OarExecutor.groovy]: again

**Comment by valleema** [comment]
OK those evil tabs have been deleted.

**Pull request merged**